Prot Warrior free_revenge to reserve_rage setting

Wyste opened the pull request:
Abilities affected: Revenge + Execute.  Simple enough change; remove the free_revenge toggle setting in favor of a setting that allows the user to reserve a set amount of rage to maintain for use on rotation defensives (Shield Block and Ignore Pain). 

35 is the default, due to Ignore Pain's cost being higher then Shield Block. For the UI, I decided a width of 3 was better looking due to the text's length and slider feeling easier to use, but feel free to change to suit your design intent.

I decided to leave Condemn out of the new setting due to it's upcoming scarcity with DF launch, and the ability has defensive value to it anyways.

This PR references Issue #1987 . In the issue you commented that any choice in changes here was going to result in more questions, but I feel as if this is a better solution then just gating Revenge behind it's proc. The user can change the setting down to 0 if they like to get the same effect as before, just like toggling the free_revenge off. 

Playing the phase2 primalist storm event with this new setting felt good and wasn't too greedy with how protection warrior plays currently.

The only future enhancement I can see to this setting would be to consider the current rolling rotation defensives as well, such as if IP is up and near cap, or SB is already rolling with a high duration, or the 2nd charge is ready. But that's a discussion for another day.

Wyste committed:
Prot Warrior free_revenge to reserve_rage setting

Hekili reviewed (changes requested):
I'd like to think more about this change.  I think the setting should be used to confirm that upon using Execute or Revenge, you'll be left with at least this much rage.

I'm also thinking we may want a readyTime or timeToReady function so that if Rage is going to be generated, the addon can recommend Execute or Revenge in a few seconds instead of just deciding it's not usable.  Does that make sense?

Wyste commented:
> I'd like to think more about this change. I think the setting should be used to confirm that upon using Execute or Revenge, you'll be left with at least this much rage.
> 
> I'm also thinking we may want a readyTime or timeToReady function so that if Rage is going to be generated, the addon can recommend Execute or Revenge in a few seconds instead of just deciding it's not usable. Does that make sense?

Yup, makes sense, and I actually would prefer it your way.

Hekili commented:
Thanks.  I'm testing a quick implementation now.  This is something that we have to watch carefully; if there are new buffs that cause a free Execute, for example, we have to make sure `timeToReady` is updated to account for it as well.  Execute is a pretty unwieldy action model at this point.

This also revealed that I've implemented the same thing two different ways over time; `timeToReady` and `readyTime` are effectively the exact same function.

Hekili commented:
Do we want to get any more elaborate?  We could ignore this when Ignore Pain is up or Shield Block is up, or when either have a certain amount of time remaining, or if they're both on CD...

I just want to ensure you can push those buttons when they're available without having been told to overspend.

Hekili commented:
See:  https://github.com/Hekili/hekili/commit/e2638c546fe2555db374ff69f7168cbbb23fdd5c

Wyste committed:
Merge branch 'dragonflight' into Prot-War-Settings-Change

Wyste committed:
bump executes usability threshold down to 20 from 40

Wyste committed:
update revenge and execute rage threshold comparison

Wyste commented:
Added some changes to bump the threshold of execute down to reserve + 20 vs reserve + 40
Updated the comparison of the threshold to be `>=` vs `>` for edge cases.

Wyste committed:
Revert "bump executes usability threshold down to 20 from 40"

This reverts commit cc945dc9cbb0f701373403fae8a64cc8713d8572.

Wyste committed:
fix formatting

Wyste reviewed (commented):
PR is now in a good state considering the changes and communication which occurred in both the PR and Discord.
